docs: improve Custom Field value propagation section (#3383)
vaadin/docs: articles/components/custom-field/index.adoc

Custom Field is optimized for wrapping the following components:

- <<../text-field#,Text Field>>
- <<../number-field#,Number Field>>
- <<../password-field#,Password Field>>
- <<../text-area#,Text Area>>
- <<../select#,Select>>
- <<../combo-box#,Combo Box>>
- <<../date-picker#,Date Picker>>
- <<../time-picker#,Time Picker>>

It can also be used to give a label, helper, and other field features for components that don't have them built-in, such as <<../list-box#,List Box>>.

== Value Type & Format

The type, format, and further propagation of the Custom Field value are handled differently in Java, and in Lit and React. The following two sections explain how you might configure these aspects depending on the programming language or approach you prefer.

ifdef::flow[]
=== Java

Custom Field is a generic class that accepts a value type. The value type can be anything: String, List, a bean, etc.

> The value propagation between Custom Field and its child components heavily depends on its structure, contained components, and their types. For this reason, no default implementation is provided. To set up the value propagation, Custom Field provides two methods that should be implemented manually: `generateModelValue`; and `setPresentationValue`.

Author: ```suggestion
When the type is specified, you need to establish how the Custom Field's value should propagate to the child components and vice versa. The value propagation heavily depends on the Custom Field structure, contained components, and their types. For this reason, no default implementation is provided. Instead, Custom Field provides two methods that should be implemented manually: `generateModelValue`; and `setPresentationValue`.
```

It can also be used to give a label, helper, and other field features for components that don't have them built-in, such as <<../list-box#,List Box>>.

> == Value Propagation & Format

Author: ```suggestion
== Value Type & Format
```

ifdef::flow[]
=== Java

Custom Field is a generic class that requires a value type. The value type can be anything: String, List, a bean, etc.

> Since Custom Field can be organized in many different ways, there is no default strategy for the value propagation between the Custom Field and its child components. Instead, Custom Field provides two methods that should be implemented manually: `generateModelValue` and `setPresentationValue`.

Author: ```suggestion
The value propagation between Custom Field and its child components heavily depends on its structure and components it contains. For this reason, no default implementation is provided. To set up the value propagation, Custom Field provides two methods that should be implemented manually: `generateModelValue` and `setPresentationValue`.
```

> Custom Field does not offer a default strategy for value propagation between the Custom Field and its child components, as the child components can vary widely in type and ways of setting their value. Instead, it provides two methods that should be implemented manually: `generateModelValue` and `setPresentationValue`.

Author: ```suggestion
Since Custom Field can be organized in many different ways, there is no default strategy for the value propagation between the Custom Field and its child components. Instead, Custom Field provides two methods that should be implemented manually: `generateModelValue` and `setPresentationValue`.
```

Custom Field is a generic class that requires a value type. The value type can be anything: String, List, a bean, etc.

> The value propagation between Custom Field and its child components heavily depends on its structure and components it contains. For this reason, no default implementation is provided. To set up the value propagation, Custom Field provides two methods that should be implemented manually: `generateModelValue` and `setPresentationValue`.

Author: ```suggestion
The value propagation between Custom Field and its child components heavily depends on its structure, contained components, and their types. For this reason, no default implementation is provided. To set up the value propagation, Custom Field provides two methods that should be implemented manually: `generateModelValue` and `setPresentationValue`.
```

=== Lit and React

> Custom Field only supports string values, but it allows customizing the format of the value with a parser and formatter. The parser and formatter define how Custom Field's value should be split to child component values and vice versa. When a child component emits a change event or Custom Field's value changes programmatically, Custom Field automatically propagates values based on the result of the parser or formatter.

Author: ```suggestion
Custom Field only supports string values, but it provides control over the format of the value with a parser and formatter. The parser and formatter define how Custom Field's value should be split to child component values and vice versa. When a child component emits a change event or Custom Field's value changes programmatically, Custom Field automatically propagates values based on the result of the parser or formatter.
```

> The default value format is a string concatenation of the child component values, separated by the `\t` character.

Reviewer: Is this a bit redundant as it's also mentioned in the _formatter_ section?

Author: Do you mean the sentence about the default format or the first one too?

Reviewer: Sorry, sentence about default format only.

Author: Good point, removed.

=== Lit and React

> Custom Field includes a parser and formatter to manage the value format. Custom Field only supports string values.

Author: ```suggestion
Custom Field only supports string values, but it allows customizing the format of the value by providing a parser and formatter.
```

> Custom Field only supports string values, but it allows customizing the format of the value by providing a parser and formatter.

Author: ```suggestion
Custom Field only supports string values, but it allows customizing the format of the value by providing a parser and formatter. The default value format is a string concatenation of the child component values, separated by the `\t` character.
```
